### ISSUE TITLE
Validator ignores format:date-time

### ISSUE THREAD

**Issue by Antares42** [everit-org/json-schema#22]
When validating against this JSON schema

	{
	    "type":"object",
	    "$schema":"http://json-schema.org/draft-04/schema#",
	        "date":{
	            "type":"string",
	            "format":"date-time"
	        }
	    }
	}

sending any String as the ``date`` will do. I've read somewhere that support for ``format`` is optional, is this the case for this validator?

**Comment by Antares42** [comment]
I can fork and create a pull request if you like.

**Comment by erosb** [comment]
@Antares42 yes, supporting the "format" keyword is optional and this validator currently does not support it. But you are not the first user asking for it, so we may add it. I'd be thankful if you could create a PR for it.

Thanks.

**Comment by Antares42** [comment]
I shouldn't be worried that my PR is flagged red as "tests failed" by Travis, right? Because that's the entire point. :-)

**Comment by erosb** [comment]
@Antares42 I've made some investigations about this issue. The json schema spec mandates an implementations supporting the "format" keyword to support the following values:

* date-time: this seems to be simple to implement
* ipv4 and ipv6: I'll take the same approach as the fge/json-schema-validator library does (it uses guava)
* hostname and uri: I'll look after these
* email: looks tricky, because email address validation is not as straightforward as it seems to be. I've tried to follow the way how fge/json-schema-validator works, but it doesn't seem to be perfect. For example it accepts somebody.@somewhere.com, while it is not a valid email address (the local part shouldn't end with a period). I could find a decent test suite in (this library)[http://isemail.info/] , I will use these tests to evaluate some java libraries if they work as expected. 


To sum up, this issue is more complicated than I expected, so I can't promise any ETA.

**Comment by Antares42** [comment]
I understand. I think we use the Apache Commons email validator in our projects, which appears to conform pretty closely to the RFC.

**Comment by rcillo** [comment]
Maybe it would be easier to implement something like tv4 `addFormat(format, validationFunction)`. This way one could provide formats as a separate library.

I think it would require a lower commitment from the library contributors and maintainers and provide a more powerful extension point.

What do you think?

**Comment by erosb** [comment]
Yes, I also thought about providing a way for users to implement their own format validators. By the way currently the built-in formats are supported (see [this branch](https://github.com/erosb/json-schema-validator/tree/format) ) but I won't have time to finish it until the middle of April, so 1.2.0 will be released in a month (hopefully).

**Comment by rcillo** [comment]
When you have the time to merge it I'll definitely use it. Great job with the library :+1:

**Comment by erosb** [comment]
I've just finished and merged "format" keyword support (and [added docs to readme](https://github.com/everit-org/json-schema#format-validators) ), tagged as version 1.2.0 .
@balazs-zsoldos please release version 1.2.0 on maven central.
Thank you.

**Comment by rcillo** [comment]
I tested the release 1.2.0 and it seems that the format for date-time is not strict. For example, the following string would be accepted `"1996-60-15T16:39:57-08:00"` or even `"1996-60-999T16:39:57-08:00"`.

I took a look at the docs for SimpleDateFormat and it seems that it's lenient by default. But I'm afraid that the RFC doesn't allow such heuristics to take place.

Maybe the code should be changed to `setLenient(false)`?

**Comment by balazs-zsoldos** [comment]
@erosb Shall I upload the release to maven-central? I am asking back because of the comment of @rcillo.

**Comment by erosb** [comment]
@balazs-zsoldos now yes, please, I've just fixed it and re-released 1.2.0 on github
@rcillo thanks for the fast feedback

**Comment by balazs-zsoldos** [comment]
@erosb Released to maven central
